### PR TITLE
F first version

### PR DESCRIPTION
**Pull request by shadowmanos**
Reimplementation of the main loop of pr2_ethercat in a new package named ros_ethercat. code relevant to ronex and the hand need to be modified accordingly.
- The launch file needs to launch ros_ethercat instead of pr2_ethercat
- in controller_plugins.xml -   base_class_type="controller_interface::ControllerBase"
- add ros_ethercat and ros_control packages as needed

a lot of code has been copied from pr2_ethercat

this code has only been tested with ronex at this time

### CODE REVIEW

**Comment by ugocupcic** [comment]
Do you think it'd be possible to show this as a diff against standard pr2_etherCAT code? This repo should be a fork of pr2_etherCAT no? (not sure if we can have a fork with a different name :S ).

**Comment by shadowmanos** [comment]
For the ros_ethercat repo
I copied and then modified main.cpp from pr2_ethercat of pr2_robot repo
https://github.com/PR2/pr2_robot/tree/hydro-devel/pr2_ethercat
and controller_spec.h from the pr2_controller_manager package of
pr2_mechanism repo
https://github.com/PR2/pr2_mechanism/tree/hydro-devel/pr2_controller_manager/include/pr2_controller_manager
.

Then added files ros_ethercat header and source.
- I believe a single package/folder like pr2_ethercat cannot be forked in
  github only complete repos.
- I avoided forking the whole pr2_robot repo because it has a lot of code
  specific to the pr2 hardware that we don't need.
- I preferred to copy the (rather small) controller_spec.h to avoid keeping
  pr2_controller_manager as a dependency.
- The main.cpp files in ros_ethercat and pr2_ethercat can be diffed if
  someone wants to see what I modified.
- Some functions of pr2_controller_manager were copied to the ros_ethercat
  class (without modifications).
  These have to do with publishing sensor_msgs::JointState and
  pr2_mechanism_msgs::MechanismStatistics messages
  that were published from pr2_controller_manager. The new controller manager
  doesn't publish anything as this responsibility has passed to the
  controllers themselves.

On 9 January 2014 07:35, Ugo Cupcic notifications@github.com wrote:

> Do you think it'd be possible to show this as a diff against standard
> pr2_etherCAT code? This repo should be a fork of pr2_etherCAT no? (not sure
> if we can have a fork with a different name :S ).
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/shadow-robot/ros_ethercat/pull/1#issuecomment-31908509
> .

## 

Manos Nikolaidis _Software Developer_
+44 (0)207 700 2487 manos@shadowrobot.com www.shadowrobot.com

## 

---

_Shadow Robot Company Ltd._
251 Liverpool Road, N1 1LX, UK

## Registered Number 3308007 (England & Wales)

**Comment by ugocupcic** [comment]
Had a lengthy chat with @markpitchless :) OK. Let's do it like this.

**Comment by markpitchless** [comment]
@shadowmanos please have quick look at the new joint_state_controller, starting this in the controller manager should be enough for it to publish joint states. You can then just remove the code you added for that. 

In GSC the controller has config like so:

``` yaml
  joint_state_controller:
    type: joint_state_controller/JointStateController
    publish_rate: 50
```

Can do that in this branch if quick (then less code to review ;-) or make a task if you think it needs more time.

https://github.com/ros-controls/ros_controllers/tree/hydro-devel/joint_state_controller

Is there a replacement for the pr2_mechanism_msgs::MechanismStatistics stuff under ros_control? Seems like we shouldn't be publishing pr2 messages in this code.

File: src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+  {
+    if (ros::param::get(g_options.rosparam_, g_robot_desc))
+    {
+      xml.Parse(g_robot_desc.c_str());
+    }
+    else
+    {
+      ROS_FATAL("Could not load the xml from parameter server: %s", g_options.rosparam_);
+      rv = -1;
+      publisher.stop(); delete rtpublisher; ros::shutdown(); return (void *)rv;
```
Comment (ugocupcic) [inline src/main.cpp]: Could this be replaced by a simple shutdown function? Used a lot below.

File: src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+  thread_param.sched_priority = sched_get_priority_max(policy);
+  pthread_setschedparam(pthread_self(), policy, &thread_param);
+
+  struct timespec tick;
+  clock_gettime(CLOCK_REALTIME, &tick);
+  period = 1e+6; // 1 ms in nanoseconds
+  ros::Duration durp(period);
+
+  // Snap to the nearest second
+  //tick.tv_sec = tick.tv_sec;
```
Comment (ugocupcic) [inline src/main.cpp]: what's happening with the sec/nsec computation? Why is sec commented out?

File: include/ros_ethercat/ros_ethercat.hpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
+/*
+ * sr_ethercat_interface.hpp
```
Comment (markpitchless) [inline include/ros_ethercat/ros_ethercat.hpp]: ros_ethercat.hpp

File: include/ros_ethercat/ros_ethercat.hpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+#include <ethercat_hardware/ethercat_hardware.h>
+#include <std_srvs/Empty.h>
+
+class ros_ethercat : public hardware_interface::RobotHW
+{
+public:
+  ros_ethercat(pr2_hardware_interface::HardwareInterface *hw, ros::NodeHandle nh) :
+    model_(hw),
+    state_(NULL),
+    cm_node_(nh, "ronex_controller_manager"),
```
Comment (markpitchless) [inline include/ros_ethercat/ros_ethercat.hpp]: I'm guessing this should be "controller_manager", ie not ronex?

File: src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+      {
+	g_stats.last_overrun = 1000;
+	g_stats.last_severe_overrun = 1000;
+      }
+      // check for overruns
+      if (g_stats.recent_overruns > 10)
+	g_stats.last_severe_overrun = 0;
+      g_stats.last_overrun = 0;
+
+      ++g_stats.overruns;
```
Comment (ugocupcic) [inline src/main.cpp:480]: ahah you're crazier than me ;)

File: include/ros_ethercat/ros_ethercat.hpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    pub_joint_state_(nh, "joint_states", 1),
```
Comment (markpitchless) [inline include/ros_ethercat/ros_ethercat.hpp:66]: Interesting, never thought to use initialisation lists for publishers!

File: src/main.cpp (hunk 1 above)
Comment (ugocupcic) [inline src/main.cpp]: never mind it's better than the jump / end

**Comment by ugocupcic** [comment]
replace all pr2 / ros (a few in the comments names etc..)

File: package.xml [hunk 7]
```diff
@@ -0,0 +1,48 @@
+<package>
+  <name>ros_ethercat</name>
+  <version>0.0.1</version>
+  <description>Main loop that runs shadow hardwarethe robot.</description>
```
Comment (markpitchless) [inline package.xml]: Maybe something like - "Main loop to run EtherCAT robot hardware."
It is not shadow specific.

File: package.xml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  <maintainer email="manos@shadowrobot.com">Manos Nikolaidis</maintainer>
```
Comment (markpitchless) [inline package.xml]: I think shadow robot team for the maintainer, like we have being doing in ronex, dms etc.

File: src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,785 @@
...
+  size_t rv = 0;
+  double last_published, last_loop_start;
+  int period;
+  int policy;
+  TiXmlElement *root;
+  TiXmlElement *root_element;
+
+  ros::NodeHandle node(name);
+
+  RealtimePublisher<diagnostic_msgs::DiagnosticArray> publisher(node, "/diagnostics", 2);
```
Comment (markpitchless) [inline src/main.cpp:287]: Should this really be forced into the root namespace, ie maybe should be "diagnostics" without the slash?

File: src/main.cpp (hunk 2 above)
Comment (shadowmanos) [inline src/main.cpp]: that statement in pr2_ethercat simply assigned a variable to itself which does nothing and the compiler complained about it. I was actually planning to delete this line eventually which I forgot to do ;-)

File: include/ros_ethercat/ros_ethercat.hpp (hunk 4 above)
Comment (shadowmanos) [inline include/ros_ethercat/ros_ethercat.hpp]: you guessed correctly! will be fixed in next commit

File: src/main.cpp (hunk 5 above)
Comment (shadowmanos) [inline src/main.cpp:480]: I'll take that as a compliment :-)

File: src/main.cpp (hunk 9 above)
Comment (shadowmanos) [inline src/main.cpp:287]: It was like this at pr2_ethercat and apparently some of our code and other packages expect it to be published there. E.g. diagnostics package at https://github.com/ros/diagnostics will subscribe to /diagnostics and publish at /diagnostics_agg that we subscribe to in our hand code.

Comment (ugocupcic) [inline src/main.cpp:287]: Yes all diagnosis are published on the same topic which can then be parsed
and analysed by different aggregators
On Jan 9, 2014 5:18 PM, "Manos Nikolaidis" notifications@github.com wrote:

> In src/main.cpp:
> 
> > -  unsigned length_;
> > -  double *history_;
> >   +};
> >   +
> >   +void *controlLoop(void *)
> >   +{
> > -  size_t rv = 0;
> > -  double last_published, last_loop_start;
> > -  int period;
> > -  int policy;
> > -  TiXmlElement *root;
> > -  TiXmlElement *root_element;
> >   +
> > -  ros::NodeHandle node(name);
> >   +
> > -  RealtimePublisher<diagnostic_msgs::DiagnosticArray> publisher(node, "/diagnostics", 2);
> 
> It was like this at pr2_ethercat and apparently some of our code and other
> packages expect it to be published there. E.g. diagnostics package at
> https://github.com/ros/diagnostics will subscribe to /diagnostics and
> publish at /diagnostics_agg that we subscribe to in our hand code.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/shadow-robot/ros_ethercat/pull/1/files#r8761028
> .

## 

---

_Shadow Robot Company Ltd._
251 Liverpool Road, N1 1LX, UK

## Registered Number 3308007 (England & Wales)

**Comment by ugocupcic** [comment]
Did you write a wiki page or something to give a bit of hindsight on that work? I think you mentioned it during a conversation?

**Comment by shadowmanos** [comment]
@ugocupcic yes I did mention in a conversation that I would write a wiki. Haven't got around to it yet. I would like to start after I modify to use joint_state_controller as requested by you and Mark. That changes what the wiki will contain.

**Comment by ugocupcic** [comment]
OK fine with me. Do we wait for that change before the merge then? Or is
this going to be addressed in a separate merge review?

On 10 January 2014 11:40, Manos Nikolaidis notifications@github.com wrote:

> @ugocupcic https://github.com/ugocupcic yes I did mention in a
> conversation that I would write a wiki. Haven't got around to it yet. I
> would like to start after I modify to use joint_state_controller as
> requested by you and Mark. That changes what the wiki will contain.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/shadow-robot/ros_ethercat/pull/1#issuecomment-32017532
> .

## 

Ugo Cupcic _Senior Software Engineer_+44 20 7700 2487ugo@shadowrobot.com
www.shadowrobot.com

## 

---

_Shadow Robot Company Ltd._
251 Liverpool Road, N1 1LX, UK

## Registered Number 3308007 (England & Wales)

**Comment by shadowmanos** [comment]
@markpitchless && @ugocupcic 
agreed I will use joint_state_controller to publish join states. This way, there will be no mysterious JointState publication from ronex! I am going make a new branch for this change since the current branch has already been tested as it is. Changes requested in this pull request have been taken care of with my last commit so feel free to merge it.

**Pull request merged**